File: ProblemSolve/147.cs
/***********************
 * Insertion Sort List *
 ***********************/

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */

public class Solution {
    public ListNode InsertionSortList(ListNode head) {
        ListNode ans = new ListNode();
        
        //하나씩 확인
        //head에서 확인할 수를 하나 고르고, ans를 순회하면서 들어갈 자리를 찾는다
        while(head != null){
            ListNode headPointer = head;
            ListNode ansPointer = ans;
            head = head.next;
            
            while(ansPointer.next != null && ansPointer.next.val < headPointer.val){
                ansPointer = ansPointer.next;
            }
            
            headPointer.next = ansPointer.next;
            ansPointer.next = headPointer;
        }
        
        return ans.next;
    }
}

File: ProblemSolve/147.cpp
/***********************
 * Insertion Sort List *
 ***********************/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* insertionSortList(ListNode* head) {
        ListNode* ans = new ListNode(0);
        
        while(head != NULL){
            ListNode* headPointer = head;
            ListNode* ansPointer = ans;
            head = head -> next;
            
            //들어갈 자리 찾기.
            while(ansPointer -> next != NULL && ansPointer -> next -> val < headPointer -> val){
                ansPointer = ansPointer -> next;
            }
            
            //사용한 node 삭제
            headPointer -> next = ansPointer -> next;
            ansPointer -> next = headPointer;
        }
        
        return ans -> next;
    }
};